### PR TITLE
feature: multi-version support for UI5 XMLView code completion

### LINKED ISSUE #468: Outdatet SAP UI5 Version 1.71.14 in @ui5-language-assistant/language-server Error
Hello,
after adding a new ..view.xml file the Code Completion for the xml assistant throws following errors.

The 1.71.14 is not longer supported by SAP.

image

Can i manually change the used ui5 version or  must it be fixed by SAP?

Thanks

Best regards
Jörn

### PR DESCRIPTION
**Pull request by petermuessig**
This PR enhances the UI5 language assistant to support multiple UI5 versions for the `XMLView` code completion. The change includes the following changes: 

* Determine the UI5 version from the workspace root `ui5.yaml` or `package.json` (fallback)
* Fallback to `1.71.49` if the version can be determined
* Make the code completion framework aware: OpenUI5, SAPUI5
* Update URLs to use the new platform agnostic short URLs (OpenUI5: https://sdk.openui5.org - SAPUI5: https://ui5.sap.com)
* Update tests to use the latest LTS versions of UI5: `1.71.49`, `1.84.27`, `1.96.11`, as well as the latest supported version `1.105.0` and the OOM version `1.60.44` for testing purposes
* Improve the JSON schema for the `api.json` to support all versions up to `1.105.0`
* Finally, update all tests to use `1.71.49` as the default version (which is the longest running LTS)

Fixes #468

### CODE REVIEW

**Comment by petermuessig** [comment]
Finally, I made the change - improving the tests was very expensive, but I'm sure this pays off! :-) 

@bd82: can please take a look into this PR? With that PR, the language assistant is now multi-version aware and reads the configuration from the `ui5.yaml` or the `package.json` in the workspace root. A next iteration of the language assistant could be to support mono-repos properly by reading the closest `ui5.yaml` or `package.json` next to the `XMLView`. But this would require some bigger rework of the semantic model usage and the extension. Let's keep that separate. In mono repos, customers can add a configuration like that to select the proper UI5 version:

`package.json` (in workspace root): 
```json
{
  [...]
  "ui5": {
    "framework": {
      "name": "OpenUI5|SAPUI5",
      "version": "1.105.0"
    },
    [...]
  },
  [...]
}
```

File: packages/language-server/package.json [hunk 1]
```diff
@@ -43,7 +43,8 @@
     "node-fetch": "2.6.1",
     "vscode-languageserver": "6.1.1",
     "vscode-languageserver-textdocument": "1.0.1",
-    "vscode-uri": "2.1.2"
+    "vscode-uri": "2.1.2",
+    "yaml": "^1.10.2"
```
Comment (bd82) [inline packages/language-server/package.json:47]: used fixed version?

Comment (bd82) [reply packages/language-server/package.json:47]: version 2.x is out

File: packages/language-server/src/ui5-model.ts [hunk 2]
```diff
@@ -11,33 +18,104 @@ import {
 } from "@ui5-language-assistant/semantic-model";
 import { Fetcher } from "../api";
 import { getLogger } from "./logger";
+import globby from "globby";
```
Comment (bd82) [inline packages/language-server/src/ui5-model.ts:21]: put imports to 3rd party libs **before** relative / mono repo imports

File: packages/language-server/package.json (hunk 1 above)
Comment (petermuessig) [reply packages/language-server/package.json:47]: I know, but this doesn't work with Node 12 anymore and you need to disable your Circle CI test, ok?

File: packages/language-server/src/ui5-model.ts (hunk 2 above)
Comment (petermuessig) [reply packages/language-server/src/ui5-model.ts]: Done

File: packages/language-server/package.json (hunk 1 above)
Comment (petermuessig) [reply packages/language-server/package.json]: Using fix version now `1.10.2` which is the last version supporting Node 12

File: packages/language-server/src/ui5-model.ts [hunk 3]
```diff
@@ -11,33 +18,104 @@ import {
...
-  modelCachePath: string | undefined
+  modelCachePath: string | undefined,
+  workspacePath: string | undefined
 ): Promise<UI5SemanticModel> {
-  return getSemanticModelWithFetcher(fetch, modelCachePath);
+  return getSemanticModelWithFetcher(fetch, modelCachePath, workspacePath);
 }
 
 // This function is exported for testing purposes (using a mock fetcher)
 export async function getSemanticModelWithFetcher(
```
Comment (bd82) [inline packages/language-server/src/ui5-model.ts:40]: can some logic be extracted? this function is now 140 lines long...

File: packages/language-server/src/ui5-model.ts [hunk 4]
```diff
@@ -170,58 +267,24 @@ function getTypeNameFix(): TypeNameFix {
...
+  cdnBaseUrl: string
+): Promise<string[]> {
+  let libs = (await readFromCache(cacheFilePath)) as string[];
+  if (libs === undefined) {
+    const url = `${cdnBaseUrl}resources/sap-ui-version.json`;
+    const response = await fetch(url);
+    if (response.ok) {
+      const versionInfo = await response.json();
+      // read libraries from version information
+      libs = versionInfo?.libraries?.map((lib) => {
```
Comment (bd82) [inline packages/language-server/src/ui5-model.ts:281]: The API.json is in a state where all the libs there are fully valid?
No filtering / exclusion needed?

File: packages/language-server/test/snapshots/xml-view-diagnostics/snapshots-utils.ts [hunk 5]
```diff
@@ -93,7 +93,8 @@ export function readSnapshotDiagnosticsLSPResponse(
 }
 
 const ui5ModelPromise = generateModel({
-  version: "1.71.14",
+  framework: "sapui5",
```
Comment (bd82) [inline packages/language-server/test/snapshots/xml-view-diagnostics/snapshots-utils.ts:96]: longer term we may want to extract this "MAGIC" string of the version if it reused in all/most tests.

File: packages/language-server/src/ui5-model.ts (hunk 3 above)
Comment (petermuessig) [reply packages/language-server/src/ui5-model.ts:40]: Yes, but I would do so with the next iteration of the language assistant making it work for mono-repos

File: packages/language-server/src/ui5-model.ts (hunk 4 above)
Comment (petermuessig) [reply packages/language-server/src/ui5-model.ts:281]: All invalid libs will not return an api.json which is fine IMO - we can implement another feature which adds an indicator in the sap-ui-version.json to know whether api.json is available or not. I would keep this for a next iteration...

File: packages/language-server/test/snapshots/xml-view-diagnostics/snapshots-utils.ts (hunk 5 above)
Comment (petermuessig) [reply packages/language-server/test/snapshots/xml-view-diagnostics/snapshots-utils.ts:96]: Sure, we would in general need to think about a more dynamic handling of the versions.

File: test-packages/test-utils/src/utils/download-ui5-resources.ts [hunk 6]
```diff
@@ -17,29 +17,32 @@ async function getLibs(version: TestModelVersion): Promise<string[]> {
   // For now we assume the libraries are the same in all versions, when we support newer versions we should
   // do a better check here
   let versionInMetadataURL: string = version;
-  if (versionInMetadataURL !== "1.76.0") {
-    versionInMetadataURL = "1.76.0";
+  if (versionInMetadataURL !== "1.71.49") {
+    versionInMetadataURL = "1.71.49";
   }
-  const url = `https://unpkg.com/@sapui5/distribution-metadata@${versionInMetadataURL}/metadata.json`;
+  //const url = `https://unpkg.com/@sapui5/distribution-metadata@${versionInMetadataURL}/metadata.json`;
```
Comment (bd82) [inline test-packages/test-utils/src/utils/download-ui5-resources.ts:23]: remove commented out "old" line?

File: test-packages/test-utils/src/utils/semantic-model-provider.ts [hunk 7]
```diff
@@ -253,29 +349,58 @@ const libraryFixes: Record<TestModelVersion, Record<string, LibraryFix[]>> = {
...
+            });
+          }
+        });
       },
     ],
-    "sap.ushell": [
+  },
+  "1.105.0": {
+    "sap.ui.mdc": [
       (content: Json): void => {
```
Comment (bd82) [inline test-packages/test-utils/src/utils/semantic-model-provider.ts:381]: is this copy / pasta? extract to a utility?

**Review by bd82 — APPROVED** [review]

File: test-packages/test-utils/src/utils/semantic-model-provider.ts (hunk 7 above)
Comment (petermuessig) [reply test-packages/test-utils/src/utils/semantic-model-provider.ts:381]: I prefer to keep the whole version `fixes` and `libraryFixes` atomic as ideally, they should be cleaned-up in the original codebase. I will try to incorporate that checks into the UI5 api.json generation to ensure that these issues will disappear and hopefully no fixes and libraryFixes will be needed anymore.

File: test-packages/test-utils/src/utils/download-ui5-resources.ts (hunk 6 above)
Comment (petermuessig) [reply test-packages/test-utils/src/utils/download-ui5-resources.ts:23]: ACK - but I like so much to keep those links to remember me... 😄 - will remove it!

**Pull request merged**